# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/public/CompileOptions.h
# Commit: 36780658b33a
# Full Hash: 36780658b33a050e82385161eab2e5e2e5e90064
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2024-12-05 09:58:04
# Regressor Bug: 1933681
# File Overlap Count: 1
# Description:
#   Bug 1933681 - Part 2: Rewrite concurrent delazification's input/output based on InitialStencilAndDelazifications. r=nbp
#   
#   The concurrent delazification internal is still using CompilationStencilMerger.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D230388
# ==============================================================================

diff -r 31d3f2901db9 -r 36780658b33a js/public/CompileOptions.h
--- a/js/public/CompileOptions.h	Wed Dec 04 23:44:33 2024 +0000
+++ b/js/public/CompileOptions.h	Wed Dec 04 23:44:33 2024 +0000
@@ -768,17 +768,22 @@
   bool hideScriptFromDebugger = false;
   bool deferDebugMetadata = false;
 
+  DelazificationOption eagerDelazificationStrategy_ =
+      DelazificationOption::OnDemandOnly;
+
   InstantiateOptions() = default;
 
   explicit InstantiateOptions(const ReadOnlyCompileOptions& options)
       : skipFilenameValidation(options.skipFilenameValidation_),
         hideScriptFromDebugger(options.hideScriptFromDebugger_),
-        deferDebugMetadata(options.deferDebugMetadata_) {}
+        deferDebugMetadata(options.deferDebugMetadata_),
+        eagerDelazificationStrategy_(options.eagerDelazificationStrategy()) {}
 
   void copyTo(CompileOptions& options) const {
     options.skipFilenameValidation_ = skipFilenameValidation;
     options.hideScriptFromDebugger_ = hideScriptFromDebugger;
     options.deferDebugMetadata_ = deferDebugMetadata;
+    options.setEagerDelazificationStrategy(eagerDelazificationStrategy_);
   }
 
   bool hideFromNewScriptInitial() const {
@@ -794,6 +799,29 @@
     MOZ_ASSERT(skipFilenameValidation == false);
     MOZ_ASSERT(hideScriptFromDebugger == false);
     MOZ_ASSERT(deferDebugMetadata == false);
+    MOZ_ASSERT(eagerDelazificationStrategy_ ==
+               DelazificationOption::OnDemandOnly);
+  }
+
+  // Assert that all fields have values compatible with the default value.
+  //
+  // This can be used in the same way as assertDefault, in case the
+  // setForceFullParse() is used on the original compile options.
+  void assertCompatibleWithDefault() const {
+    MOZ_ASSERT(skipFilenameValidation == false);
+    MOZ_ASSERT(hideScriptFromDebugger == false);
+    MOZ_ASSERT(deferDebugMetadata == false);
+
+    // The instantiation step uses the eagerDelazificationStrategy_ field
+    // only for TransitiveCompileOptions::populateDelazificationCache().
+    //
+    // Both the default OnDemandOnly and
+    // the ParseEverythingEagerly from setForceFullParse() returns
+    // false, and they're are compatible.
+    MOZ_ASSERT(eagerDelazificationStrategy_ ==
+                   DelazificationOption::OnDemandOnly ||
+               eagerDelazificationStrategy_ ==
+                   DelazificationOption::ParseEverythingEagerly);
   }
 #endif
 };